fish-rust/src/builtins/ulimit.rs

```diff
@@ -1,16 +1,468 @@
+        wopt(
+            L!("virtual-memory-size"),
+            woption_argument_t::no_argument,
+            'v',
+        ),
+        wopt(L!("swap-size"), woption_argument_t::no_argument, 'w'),
+        wopt(L!("realtime-maxtime"), woption_argument_t::no_argument, 'y'),
+        wopt(L!("kernel-queues"), woption_argument_t::no_argument, 'K'),
+        wopt(L!("ptys"), woption_argument_t::no_argument, 'P'),
+        wopt(L!("threads"), woption_argument_t::no_argument, 'T'),
+        wopt(L!("help"), woption_argument_t::no_argument, 'h'),
+    ];
+
+    let mut opts = Options::default();
+
+    let mut shim_args: Vec<&wstr> = args.to_vec();
+    let mut w = wgetopter_t::new(SHORT_OPTS, LONG_OPTS, &mut shim_args);
+
+    while let Some(c) = w.wgetopt_long() {
+        match c {
+            'a' => opts.report_all = true,
+            'H' => opts.hard = true,
+            'S' => opts.soft = true,
+            'b' => opts.what = RLIMIT_SBSIZE(),
+            'c' => opts.what = RLIMIT_CORE(),
+            'd' => opts.what = RLIMIT_DATA(),
+            'e' => opts.what = RLIMIT_NICE(),
+            'f' => opts.what = RLIMIT_FSIZE(),
+            'i' => opts.what = RLIMIT_SIGPENDING(),
+            'l' => opts.what = RLIMIT_MEMLOCK(),
+            'm' => opts.what = RLIMIT_RSS(),
+            'n' => opts.what = RLIMIT_NOFILE(),
+            'q' => opts.what = RLIMIT_MSGQUEUE(),
+            'r' => opts.what = RLIMIT_RTPRIO(),
+            's' => opts.what = RLIMIT_STACK(),
+            't' => opts.what = RLIMIT_CPU(),
+            'u' => opts.what = RLIMIT_NPROC(),
+            'v' => opts.what = RLIMIT_AS(),
+            'w' => opts.what = RLIMIT_SWAP(),
+            'y' => opts.what = RLIMIT_RTTIME(),
+            'K' => opts.what = RLIMIT_KQUEUES(),
+            'P' => opts.what = RLIMIT_NPTS(),
+            'T' => opts.what = RLIMIT_NTHR(),
+            'h' => {
+                builtin_print_help(parser, streams, cmd);
+                return STATUS_CMD_OK;
+            }
+            ':' => {
+                builtin_missing_argument(parser, streams, cmd, args[w.woptind - 1], true);
+                return STATUS_INVALID_ARGS;
+            }
+            '?' => {
+                builtin_unknown_option(parser, streams, cmd, args[w.woptind - 1], true);
+                return STATUS_INVALID_ARGS;
+            }
+            _ => {
+                panic!("unexpected retval from wgetopt_long");
+            }
+        }
+    }
+
+    if opts.report_all {
+        print_all(opts.hard, streams);
+    }
+
+    if opts.what == RLIMIT_UNKNOWN() {
+        streams.err.append(wgettext_fmt!(
+            "%ls: Resource limit not available on this operating system\n",
+            cmd
+        ));
+
+        builtin_print_error_trailer(parser, streams.err, cmd);
+        return STATUS_INVALID_ARGS;
+    }
+
+    let what = opts.what as c_uint; // Only negative value is RLIMIT_UNKNOWN
```

Comment:
better to use try_from()/try_into() + unwrap() IMO.
(In future we'll hopefully use an enum.)